File: python-bindings/tests/xse_plugin_validation_conformance.py
"""Actual Address Library factories and plugin checks over disposable files."""

import os
from pathlib import Path
from tempfile import TemporaryDirectory


def observe_xse_plugins(fixture):
    """Read native metadata and compare convenience validation to the typed checker."""
    import classic_scangame as native

    with TemporaryDirectory(prefix="classic-xse-plugins-") as directory:
        root = Path(directory)
        for name, content in fixture["files"].items():
            (root / name).write_bytes(content.encode())
        previous = Path.cwd()
        try:
            # Registry fallback must use embedded data, never ambient checkout YAML.
            os.chdir(root)
            factory = {"Original": "original", "NextGen": "next_gen", "Vr": "vr"}[
                fixture["version"]
            ]
            info = getattr(native.AddressLibInfo, factory)()
            version = getattr(native.GameVersion, fixture["version"])
            checker = native.XseChecker(root, version)
            before = {
                p.name: p.read_bytes().decode() for p in root.iterdir() if p.is_file()
            }
            result = checker.check()
            message = checker.validate()
            if native.check_xse_plugins(root, version) != message:
                raise ValueError("XSE convenience validation disagrees with checker")
            return {
                "info": {
                    "version": str(info.version).split(".")[-1],
                    "filename": info.filename,
                    "description": info.description,
                    "url": info.url,
                },
                "result": str(result).split(".")[-1],
                "message": message,
                "beforeFiles": before,
                "files": {
                    p.name: p.read_bytes().decode()
                    for p in root.iterdir()
                    if p.is_file()
                },
            }
        finally:
            os.chdir(previous)

File: node-bindings/classic-node/__test__/xse_plugin_validation_conformance.ts
import {mkdtemp, readdir, readFile, rm, writeFile} from "node:fs/promises";
import {tmpdir} from "node:os";
import {join} from "node:path";
import * as classic from "../index.js";

/** Validate actual plugin filenames and native metadata using embedded registry fallback. */
export async function observeXsePlugins(fixture: Record<string, any>): Promise<Record<string, any>> {
    const root = await mkdtemp(join(tmpdir(), "classic-xse-plugins-")), previous = process.cwd();
    try {
        for (const [name, content] of Object.entries(fixture.files)) await writeFile(join(root, name), content as string);
        process.chdir(root);
        const info = classic.getAddressLibInfo(fixture.version),
            checker = new classic.JsXseChecker(root, fixture.version);
        const files = async () => Object.fromEntries(await Promise.all((await readdir(root)).map(async name => [name, await readFile(join(root, name), "utf8")])));
        const beforeFiles = await files(), result = checker.check(), message = checker.validate();
        if (classic.checkXsePlugins(root, fixture.version) !== message) throw new Error("XSE convenience validation disagrees with checker");
        return {info, result, message, beforeFiles, files: await files()};
    } finally {
        process.chdir(previous);
        await rm(root, {recursive: true, force: true});
    }
}
